heartbeat/monitors/active/http/http_test.go

```diff
@@ -0,0 +1,210 @@
+    }
+    defer resp.Body.Close()
+    copyHeader(w.Header(), resp.Header)
+    w.WriteHeader(resp.StatusCode)
+    io.Copy(w, resp.Body)
+}
+
+func transfer(destination io.WriteCloser, source io.ReadCloser) {
+    defer destination.Close()
+    defer source.Close()
+    io.Copy(destination, source)
+}
+
+func copyHeader(dst, src http.Header) {
+    for k, vv := range src {
+        for _, v := range vv {
+            dst.Add(k, v)
+        }
+    }
+}
+
+func runAndAssert(t *testing.T, rawCfg map[string]interface{}) {
+	// create an http monitor job, run it and make sure that the beat event indicates that the host is accessible
+	info := monitors.Info{}
+	cfg, _ := common.NewConfigFrom(rawCfg)
+	jobs, err := create(info, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, 1, len(jobs))
+
+	event, _, err := jobs[0].Run()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Logf("Event %v", event)
+
+	monitorStatus, err := event.GetValue("monitor.status")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	httpStatus, err := event.GetValue("http.response.status")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	hasError, err := event.HasKey("error")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, "up", monitorStatus)
+	assert.Equal(t, 200, httpStatus)
+	assert.False(t, hasError)
+}
+
+func TestSSLVerificationModeToNone(t *testing.T) {
+	serverHandler := &ServerHandler{}
+	server := httptest.NewTLSServer(serverHandler)
+	defer server.Close()
+
+	rawCfg := map[string]interface{}{
+		"urls": []string{
+			server.URL,
+		},
+		"ssl": map[string]interface{}{
+			"verification_mode": "none",
+		},
+	}
+
+	runAndAssert(t, rawCfg)
+	assert.True(t, serverHandler.Invoked)
+}
+func TestHTTPProxyForHTTPSEndpoint(t *testing.T) {
+	serverHandler := &ServerHandler{}
+	server := httptest.NewTLSServer(serverHandler)
+	defer server.Close()
+
+	proxyHandler := &ProxyHandler{}
+	proxyServer := httptest.NewServer(proxyHandler)
+	defer proxyServer.Close()
+
+	rawCfg := map[string]interface{}{
+		"urls": []string{
+			server.URL,
+		},
+		"proxy_url": proxyServer.URL,
+		"ssl": map[string]interface{}{
+			"verification_mode": "none",
+		},
+	}
+
+	runAndAssert(t, rawCfg)
+	assert.True(t, proxyHandler.Invoked)
+}
+
+func TestHTTPSProxyForHTTPSEndpoint(t *testing.T) {
+	serverHandler := &ServerHandler{}
+	server := httptest.NewTLSServer(serverHandler)
+	defer server.Close()
+
+	proxyHandler := &ProxyHandler{}
+	proxyServer := httptest.NewTLSServer(proxyHandler)
+	defer proxyServer.Close()
+
+	rawCfg := map[string]interface{}{
+		"urls": []string{
+			server.URL,
+		},
+		"proxy_url": proxyServer.URL,
+		"ssl": map[string]interface{}{
+			"verification_mode": "none",
+		},
+	}
+
+	runAndAssert(t, rawCfg)
+	assert.True(t, proxyHandler.Invoked)
+}
+
+func TestHTTPProxyConfiguredViaEnv(t *testing.T) {
+	t.Skip("This test cannot be executed until https://github.com/golang/go/issues/22079 is merged, as the proxy settings are being cached and there is no way to refresh them.")
```

Comment:
Unfortunately it not possible to test/showcase the first issue (taking the environment variables into account), as the proxy settings are being cached once by the http module and there is no way (yet) to reload them.